### PR TITLE
When an exception is not handled, show the stack trace

### PR DESCRIPTION
**Pull request by m-kuhn**
screenshot from 2017-03-30 13-31-19

### CODE REVIEW

**Comment by NathanW2** [comment]
I assume exception means application crash here?

**Comment by m-kuhn** [comment]
No, exception means geos (and other) exceptions showing up as "Unhandled exception" dialogs without any hint what has gone wrong so far.

Any chance you can implement that for windows? ;)

Application crashes would be yet another nice enhancement.

**Comment by NathanW2** [comment]
Can we put it into the message bar and just log it?

**Comment by m-kuhn** [comment]
We could, but I prefer it this way, because

* this is really severe
* this should not happen, we really did something wrong if this happens
* we therefore want to get the information where an exception could bubble up so far
* we want this information to get to us with the highest probable chance
* As an example, in the reshape scenario I had last week, the whole UI was b0rked and I had no clue why (i.e. requires restarting qgis to continue the work). I wish I had the windows implementation of this showing up back then.
* Therefore UX is really secondary here (a modal dialog is shown already now, and the heck I couldn't care less about this one)
* I repeat, it's severe :stuck_out_tongue_closed_eyes: 

Now that we all know it's very severe...

* I almost never the the current "unhandled exception" message box
* it is only shown in debug builds, which are there for testing and reporting

**Comment by nyalldawson** [comment]
I'm in favour, but instead of a giant message box could we use a custom dialog with the receipt inside a scrollable text edit? It'd be easier to read add would also allow copying of the trace for bug reports.

**Comment by m-kuhn** [comment]
that's what i hoped would happen anyway and i was just too lazy to write code for a ui that's shown only once in a year. and the text was copyable already...

...

...

ok ill do it

**Comment by NathanW2** [comment]
:)

File: src/core/qgsstacktrace.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
+/***************************************************************************
+  qgsstacktrace - %{Cpp:License:ClassName}
```
Comment (jef-n) [inline src/core/qgsstacktrace.cpp]: template leftovers?

File: src/core/qgsstacktrace.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ ***************************************************************************/
+
+#include "qgsstacktrace.h"
+#include <stdlib.h>
+#include <execinfo.h>
+#include <cxxabi.h>
+
+#ifdef QGISDEBUG
+#ifdef Q_OS_LINUX
+QStringList QgsStacktrace::trace( unsigned int maxFrames )
```
Comment (jef-n) [inline src/core/qgsstacktrace.cpp]: There's already code in [src/app/main.cpp:196](https://github.com/qgis/QGIS/blob/master/src/app/main.cpp#L196) to show backtraces (including windows).

Comment (m-kuhn) [inline src/core/qgsstacktrace.cpp]: That looks like some really powerful code there, thanks for the hint.

**Comment by m-kuhn** [comment]
Thinking out loud, since we are apparently able to get stack traces in code, would it be possible to 

* offer a window for bugreporting
* directly upload the stack to some webservice (e.g. `gist.github.com`)
* add some additional notes from the user (what have you been doing when this happened)
* search the issue tracker for existing reports with the same signature
* update the current issue or add a new issue

Similar to this plugin here:
https://github.com/lutraconsulting/qgis-report-plugin#qgis-report-plugin

**Comment by jef-n** [comment]
@m-kuhn there were thoughts by @wonder-sk to provide [breakpad](https://github.com/google/breakpad) at some point - that's why there is a symbol store for windows - but apparently that never was followed up.  There are projects using this with [cmake](https://github.com/Mendeley/breakpad) and/or [Qt](https://github.com/JPNaude/dev_notes/wiki/Using-Google-Breakpad-with-Qt)

File: src/core/qgsstacktrace.cpp (hunk 2 above)
Comment (nyalldawson) [inline src/core/qgsstacktrace.cpp]: @jef-n how do you access these backtraces on windows?

Comment (jef-n) [inline src/core/qgsstacktrace.cpp]: @nyalldawson [main.cpp:269](https://github.com/qgis/QGIS/blob/master/src/app/main.cpp#L269)

**Comment by NathanW2** [comment]
I have managed to get full stack dumps working on Windows with RelWithDebInfo.  @m-kuhn I'm keen to improve this with you once easter is over.   If these kinds of exceptions are severe I think it's best to handle the ones we can and crash out QGIS for the others showing my new crash dialog.

**Pull request closed without merging**